### ISSUE TITLE
CSS: contain-intrinsic-size: auto, and converting to a shorthand property

### ISSUE THREAD

**Issue by cbiesinger** [w3ctag/design-reviews#624]
Ya ya yawm TAG!

I'm requesting a TAG review of contain-intrinsic-size: auto, and converting to a shorthand property.

This adds an "auto" keyword to contain-intrinsic-size, which lets the rendering engine remember the last
used size so that the author does not have to guess correctly (with a fallback size if the element was
never laid out without size containment).
Also, longhand properties were added, including logical axes: contain-intrinsic-width, contain-intrinsic-height, contain-intrinsic-block-size, contain-intrinsic-inline-size

  - Explainer¹ (minimally containing user needs and example code): https://gist.github.com/cbiesinger/f2378dbcd215495c3a1daf9696a8e91f
  - Specification URL: https://drafts.csswg.org/css-sizing-4/#propdef-contain-intrinsic-size
  - Tests: not written yet; I am filing this review before implementation and test writing has started.
  - Security and Privacy self-review²: n/a
  - GitHub repo (if you prefer feedback filed there): https://github.com/w3c/csswg-drafts/issues
  - Primary contacts (and their relationship to the specification):
      - Vladimir Levin (@vmpstr), Google
      - Christian Biesinger (@cbiesinger), Google
  - Organization(s)/project(s) driving the specification: Google and the CSSWG
  - Key pieces of existing multi-stakeholder review or discussion of this specification: https://github.com/w3c/csswg-drafts/issues/5668#issuecomment-759602039
  - External status/issue trackers for this specification (publicly visible, e.g. Chrome Status): https://chromestatus.com/feature/6740477866934272

Further details:

  - [x] I have reviewed the TAG's [Web Platform Design Principles](https://w3ctag.github.io/design-principles/)
  - Relevant time constraints or deadlines: Ideally within the next month or so
  - The group where the work on this specification is currently being done: CSSWG
  - The group where standardization of this work is intended to be done (if current group is a community group or other incubation venue): CSSWG
  - Major unresolved issues with or opposition to this specification: none
  - This work is being funded by: Google

You should also know that...

[please tell us anything you think is relevant to this review]

We'd prefer the TAG provide feedback as (please delete all but the desired option):

  🐛 open issues in our GitHub repo for **each point of feedback**

**Comment by LeaVerou** [comment]
My thoughts while looking at this:

First, this definitely solves a real problem. Indeed, it's quite a hassle if authors need to guess the correct intrinsic size for this property, and it creates a maintenance nightmare when there are a lot of elements whose intrinsic sizes need to be contained. 

On the other hand, CSS has always been this nice reactive language in which things apply and then they stop applying without leaving any side effects behind. Unlike (common) programming languages, there are no ordered steps, the current state is always a function of what currently applies, and there is value in that simplicity. The concept of last remembered size that is introduced here breaks this fundamental assumption, introducing side effects that persist after properties have stopped applying. It introduces state that is not inspectable and has no obvious cause. Is there any other existing CSS feature that produces lasting side effects in that way?
Have other ways to solve this problem been sufficiently explored that do not change the way CSS works so dramatically?

**Comment by vmpstr** [comment]
Thank you for your feedback!

> The concept of last remembered size that is introduced here breaks this fundamental assumption, introducing side effects that persist after properties have stopped applying.

`contain-intrinsic-size` only has an effect if we have size containment, so the side effect here is tied to the presence of size containment (whether introduced by `contain` property directly or via something like `content-visibility`). In other words, I claim that the side effect introduced here does not persist after the property has stopped applying -- it's just not the same property, if that makes sense. But that's a bit of a pedantic point. Other than that, you're right that this introduces a state that is captured at a time different from when the property begin applying. 

> Is there any other existing CSS feature that produces lasting side effects in that way?

I'm not aware of anything in CSS that relies on previous sizing like this. However, we do have things like `content-visibility: auto` which depend on element visibility, relying essentially on `IntersectionObserver` timing. There are also things like container queries, which selectively apply different styles depending on the current state / size of elements. It's not the same as `contain-intrincis-size auto`, for sure, but I feel like it falls in a similar bucket of things that aren't exactly side-effect free.

> Have other ways to solve this problem been sufficiently explored that do not change the way CSS works so dramatically?

The piece of data that we want to capture here is the size that really only the UA knows for sure. Script, for example, can get at this size and the advice that we've been giving developers is to use a polyfill of this feature, directly using `ResizeObserver` or `getBoundingClientRect` to set appropriate values of `contain-intrinsic-size` at appropriate times. But in order for us to do this "automatically" in CSS we would need to snapshot this size before we apply size containment and use it at the time that size containment applies. So fundamentally, there is a timing dependency here that I believe is unavoidable.

**Comment by cbiesinger** [comment]
We are proposing a change to this API to reduce the "side-effects from a property that no longer applies" issue here:
https://github.com/w3c/csswg-drafts/issues/6308

If agreed to by the CSSWG, the scope of that issue will be much lower.

We consider it a huge benefit of this feature that no JavaScript code is required to get faster rendering while maintaining reasonable scrollbar behavior. It is difficult to write JavaScript code that works with arbitrary websites, because there is no good/efficient way to find elements to which `contain-intrinsic-size` was applied. It also increases the barrier for entry to use this feature, from just applying a couple of CSS properties to having to find a "polyfill" or writing code yourself.

cc @chrishtr in case he has further input.

**Comment by LeaVerou** [comment]
We looked at this again in our Gethen virtual f2f meeting.

While reducing the need for JS libraries and improving performance is definitely a goal we sympathize with, we are still concerned that this feature breaks one of the fundamental design principles of CSS, one that has been established for over twenty years. Breaking this assumption is not only confusing for authors, but has the potential to break tooling (e.g. WYSIWYG editors). While the new iteration reduces the cases where the side effects apply to when `content-visibility: auto` *also* applies, it is still introducing side effects and uninspectable state.

If side effects are truly the only way to solve this problem, have you explored solving the problem with a JS API instead?

**Comment by cbiesinger** [comment]
Thanks for the feedback!

TL;DR of the below: we think this feature is best viewed as a UA default behavior triggered by CSS, not a side-effect of CSS algorithms; there is precedent in features such as scrolling.

Full version:

After your latest comment, we thought some more about the issue of side-effects, and how to express it in some other way such as JS.

Regarding JS: as we mentioned previously, the feature can be polyfilled via a ResizeObserver, but we think it is (a) too expensive/difficult for developers (especially CSS-oriented developers) to be expected to reliably achieve, and (b) something almost all developers using content-visibility: auto would want (developer feedback from multiple partners has confirmed this).

We don’t think this should be viewed as a side-effect of a CSS algorithm (which we agree is a bad thing generally), but as a UA default behavior. Compare it with scrolling: scrolling is a super complicated UA default behavior that is triggered by CSS via `overflow:scroll`, and has lots of built-in behaviors that have visual side-effects, such as scrollbar gutters, scroll offsets and so on.

This feature should be viewed similarly - it’s a thing that developers want to do and can potentially do themselves, but there is enough efficiency/convenience/defaults-work-well value that it makes sense for the UA to provide the functionality as a built-in.

Note that the feature is carefully specified to happen at ResizeObserver timing, so there is no magic implementation advantage of the browser over script. The only magic thing about this feature is where it stores the cached intrinsic size. Conceptually it stores it on a property of Element objects (and that’s exactly what the Chromium implementation does). But we don’t propose exposing that as a JS property (analogous to `scrollTop`) to web developers at this time, nor do we propose exposing an event (analogous to `scroll`) that can allow `preventDefault`. If there were demand for such features we could expose them in the future (https://github.com/WICG/display-locking/issues/112).

**Comment by cbiesinger** [comment]
Thanks for the response, @atanassov . Are you concerned about a transition from `contain-intrinsic-size: auto 1px` to `contain-intrinsic-size: 2px`? For auto to non-auto (or vice versa), we consider those as incompatible and just flip at 50%, so it shouldn't make a difference which one we pick there. For `auto 1px` to `auto 2px` it won't matter because we'll just end up using auto in both cases.

cc @lilles @flackr in case they have any input.

**Comment by flackr** [comment]
Just to add on to this, treating auto as non-interpolable with non-auto values is the same as for other length properties which can be auto. Similarly auto -> auto is not a transition as they are the same value.

**Comment by atanassov** [comment]
Sounds reasonable and consistent with any other to/from `auto` transition. Please consider adding a note in the spec in order to make that easier to understand for authors. It wasn't obvious to me, though it was the anticipated answer knowing how the rest of CSS works :)

**Comment by atanassov** [comment]
We did another quick pass of the issue during our plenary meeting today. The overall shape of the feature and use cases addressed are a good fit to CSS and web, thus we are happy to close the review.

One remaining ask continues to be about the transitions/animation note we'd like to see added to the spec.

Thank you for working with us and good luck with finishing the work with the CSSWG.
